api/disk.go

```diff
@@ -4,32 +4,54 @@ import (
 	"fmt"
 	"log"
 	"strconv"
+	"time"
 )
 
-func (api *API) ResizeDisk(instanceID int, params map[string]interface{}) (map[string]interface{}, error) {
-	data := make(map[string]interface{})
-	failed := make(map[string]interface{})
-	id := strconv.Itoa(instanceID)
-	log.Printf("[DEBUG] go-api::disk::resize instance ID: %s", id)
-	path := fmt.Sprintf("api/instances/%s/disk", id)
+func (api *API) ResizeDisk(instanceID int, params map[string]interface{}, sleep, timeout int) (map[string]interface{}, error) {
+	var (
+		id   = strconv.Itoa(instanceID)
+		path = fmt.Sprintf("api/instances/%s/disk", id)
+	)
+	log.Printf("[DEBUG] go-api::resizeDisk::resizeDiskWithRetry path: %s, "+
+		"attempt: %d, sleep: %d, timeout: %d", path, 1, sleep, timeout)
+	return api.resizeDiskWithRetry(id, params, 1, sleep, timeout)
+}
+
+func (api *API) resizeDiskWithRetry(id string, params map[string]interface{}, attempt, sleep, timeout int) (map[string]interface{}, error) {
+	var (
+		data   = make(map[string]interface{})
+		failed = make(map[string]interface{})
+		path   = fmt.Sprintf("api/instances/%s/disk", id)
+	)
+
 	response, err := api.sling.New().Put(path).BodyJSON(params).Receive(&data, &failed)
 	if err != nil {
 		return nil, err
 	}
 
-	switch {
-	case response.StatusCode == 200:
+	switch response.StatusCode {
+	case 200:
 		if err = api.waitUntilAllNodesReady(id); err != nil {
 			return nil, err
 		}
 		return data, nil
-	case response.StatusCode == 400:
+	case 400:
 		switch {
 		case failed["error_code"] == nil:
 			break
 		case failed["error_code"].(float64) == 40002:
 			return nil, fmt.Errorf("Resize disk failed: %s", failed["error"].(string))
+		case failed["error_code"].(float64) == 40003:
+			if failed["error"] == "Timeout talking to backend" {
```

Comment:
I think we have (or soon have) introduced more codes in the backend so we don't need to check on the error message?

Reply:
Currently we don't have a specific error_code when we need to use RPC requests, just the error: "Timeout talking to backend". In this case 40003 refers to validate cluster health, which do a RPC request. Agree that we should have an global error_code for this too, to avoid doing the string compare.

Reply:
Oh I see. I think we should address that, so we can avoid adding code in the provider coupled to error messages strings in the backend.

Reply:
Ah the cluster health timeout happens when you bootstrap an instance and also use the extra disk resource. I was thinking this was polling to see when the disk resized had completed (because [this comment](https://github.com/cloudamqp/terraform-provider-cloudamqp/pull/194/files#diff-1e0846d663cf06f56b7560763fd3444201d8aae7595d2f7ced7b55ef4c098a2bR65) alludes to that)

Reply:
My bad, that needs to be corrected. Yes noticed cluster health timeout specifically during bootstrap. Could succeed but take more than our default RPC timeout (28 seconds), and then thrown as error. However only noticed this for GCE until post.bootstrap.done is finished. Cluster health is also validated every time resizing disk is invoked.

Reply:
New func that also use sleep/timeout for checking ready nodes.

As you mention before, huge difference between squirrel and bunny for GCE. Squirrel get stuck in post.bootstrap and timeout in the backend. While Bunny, done in one go.